fem/bilininteg_ext.cpp

```diff
@@ -378,294 +371,376 @@ void PADiffusionApply2D(const int NE,
 {
    const int D1D = T_D1D ? T_D1D : d1d;
    const int Q1D = T_Q1D ? T_Q1D : q1d;
-   MFEM_VERIFY(D1D <= MAX_D1D, "");
-   MFEM_VERIFY(Q1D <= MAX_Q1D, "");
-
-   const DeviceMatrix B(b, Q1D, D1D);
-   const DeviceMatrix G(g, Q1D, D1D);
-   const DeviceMatrix Bt(bt, D1D, Q1D);
-   const DeviceMatrix Gt(gt, D1D, Q1D);
+   const int NBZ = T_NBZ ? T_NBZ : 1;
+   const int MQ1 = T_Q1D ? T_Q1D : MAX_Q1D;
+   const int MD1 = T_D1D ? T_D1D : MAX_D1D;
+   MFEM_VERIFY(MQ1 == MD1, "");
+   MFEM_VERIFY(D1D <= MD1, "");
+   MFEM_VERIFY(Q1D <= MQ1, "");
+   const DeviceMatrix b(_b, Q1D, D1D);
+   const DeviceMatrix g(_g, Q1D, D1D);
+   const DeviceMatrix bt(_bt, D1D, Q1D);
+   const DeviceMatrix gt(_gt, D1D, Q1D);
    const DeviceTensor<3> op(_op, 3, Q1D*Q1D, NE);
    const DeviceTensor<3> x(_x, D1D, D1D, NE);
    DeviceTensor<3> y(_y, D1D, D1D, NE);
-
-   MFEM_FORALL(e, NE,
+   MFEM_FORALL_2D(e, NE, Q1D, Q1D, NBZ,
    {
       const int D1D = T_D1D ? T_D1D : d1d; // nvcc workaround
       const int Q1D = T_Q1D ? T_Q1D : q1d;
-
-      double grad[MAX_Q1D][MAX_Q1D][2];
-      for (int qy = 0; qy < Q1D; ++qy)
+      const int NBZ = T_NBZ ? T_NBZ : 1;
+      const int MQ1 = T_Q1D ? T_Q1D : MAX_Q1D;
+      const int MD1 = T_D1D ? T_D1D : MAX_D1D;
+      MFEM_SHARED double B[MQ1][MD1];
```

Comment:
Are these 2D arrays actually being stored as MQ*MD linear arrays with translations for the index math?

Reply:
Yes, that's how I see it too.

Reply:
I looked it up and I believe the compiler guarantees that we get contiguous memory for these multidimensional arrays, so no worries there.  Although it may be using extra shared memory to store the B[MQ1] pointers, I'm not sure.